### ISSUE TITLE
[CS2] Enhancement: Explicitly returning in statement-like (switch, try, etc) expressions

### ISSUE THREAD

**Issue by tphecca** [jashkenas/coffeescript#4612]
I'd like to be able to use the full power of IIFEs within my `switch` expression, namely the ability to return manually, without having to go out of my way to set the last statement of it to immediately return something.
Here is an example of explicitly setting the last statement to `undefined`:
```coffeescript
# within another function...
target2 = switch target = bm.shift()
  when "r", "role"
    unless bm[0]? then await sendBack("No role specified."); undefined else
      # this extra level of indentation could be avoided
      myRole = bm.shift()
      myRole = myRole.replace /[\<\>\@\&]/g, ""
      foundRole = message.guild.roles.find (role) =>
        role.id is myRole or role.name.toLowerCase().indexOf(myRole.toLowerCase()) >= 0
      unless foundRole? then await sendBack("Invalid role specified."); undefined else foundRole
      # too much code in one line, this could be return instead of undefined else
  # (many more when statements doing somewhat similar things go here)
```
I'd rather be able to do something like this:
```coffeescript
# within another function...
target2 = switch target = bm.shift()
  when "r", "role"
    unless bm[0]? then await sendBack("No role specified."); return
    # much easier for me to check for even more things like this
    # no extra level of indentation, the code stops here if bm[0] doesn't exist
    myRole = bm.shift()
    myRole = myRole.replace /[\<\>\@\&]/g, ""
    foundRole = message.guild.roles.find (role) =>
      role.id is myRole or role.name.toLowerCase().indexOf(myRole.toLowerCase()) >= 0
    return foundRole if foundRole?
    await sendBack("Invalid role specified."); return
    # I don't care about whatever sendBack resolves with (I await to catch errors),
    # so I want my variable to be undefined so I can rely on the existential operator
```
When I attempt that, I get this error:
```
error: cannot use a pure statement in an expression
      unless bm[0]? then await sendBack("No role specified."); return
                                                               ^^^^^^
```
The point of returning undefined is so that I can rely on the existential operator (`target2?`) to tell whether my operations completed, and move to the next phase of processing that array. More initial (potentially unsatisfied) checks and actions in that code would be an exponential, unreadable mess of `unless ...  then ...; undefined else ...` and indentation.

Being able to `return` within a `switch` expression (or any other statement-like expression) allows me to flatten these checks and just bypass returning if they are satisfied (as one would do with a normal function that does many checks). I prefer this over writing a separate function with a `switch` statement, because it is cleaner as an expression which automatically generates an IIFE.

To me, this would make statement-like expressions much more useful for simplifying complex routines, particularly the one in my example which involves a variable number of elements in an array (maybe at this point I should be writing a lexer for it... but it's already too similar to a lexer).

TL;DR I want to return something else instead of returning the last statement in IIFEs generated by `switch` expressions

**Comment by vendethiel** [comment]
`return` is reserved to functions. If you want to "abuse" `return` here, you can just use an explicit IIFE.

```coffee
target2 = do -> switch target = bm.shift()
   ... your code
```

**Comment by tphecca** [comment]
Oh, I thought that `do ->` IIFE syntax was only a LiveScript/Coco thing. (the more I think about LS, the more I miss it...) That works. Thanks!

**Comment by vendethiel** [comment]
> (the more I think about LS, the more I miss it...)

Use whatever works for you ;-). Coco's addition was `new ->`, not `do ->`.

**Comment by tphecca** [comment]
Oh, another thing: using that `do ->`, it closes my `target` variable within the IIFE instead of setting it in the scope of the parent function. I can just write `target = bm.shift()` on its own line, but it was neater to define it within the switch expression. I tried `do (target) ->` but then it doesn't define `target` anywhere and puts it in the IIFE's arguments (and doesn't use it).

**Comment by connec** [comment]
In CoffeeScript, new variables are always scoped to the closest enclosing function call. You can close over `target` *and* set it by assigning it a default value:

```coffee
do (target = bm.shift()) -> switch target
  when 'r', 'role'
    unless bm[0]? then await sendBack("No role specified."); return
    ...
```

You may find you can get a neater syntax for this pattern by pulling the entire switch (and possibly its branches) into individual functions:

```coffee
# within another function...
target2 = handle bm

handle = (bm) -> switch target = bm.shift()
  when 'r', 'role' then handleRole bm
  when '...'       then handleSomethingElse bm

handleRole = (bm) ->
  unless bm[0]? ...; return

# etc
```
